### DISCUSSION TITLE
Best practices for testing schedulers.

### DISCUSSION THREAD

**Discussion by hwdef** [kubernetes-sigs/kwok#981] [Q&A] [answered]
I want to test some scenarios using 3rd party scheduler like volcano. But I don't understand how to use kwok in this scenario.These are my steps:

1. install kwok
2. kwokctl create cluster --name=kwok
3. create a fake node https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/#create-a-node
4. install volcano `kubectl create -f https://raw.githubusercontent.com/volcano-sh/volcano/master/installer/volcano-development.yaml`

and I got this error 

```
NAMESPACE        NAME                                   READY   STATUS    RESTARTS   AGE
volcano-system   volcano-admission-6cd4c958c9-l2pkl     0/1     Pending   0          9m24s
volcano-system   volcano-admission-init-mhzv8           0/1     Pending   0          9m24s
volcano-system   volcano-controllers-7d7c4bb8b6-7j4qq   0/1     Pending   0          9m24s
volcano-system   volcano-scheduler-7db897f747-7lb9f     0/1     Pending   0          9m24s
```

```
Warning  FailedScheduling  4m14s  default-scheduler  0/1 nodes are available: 1 node(s) had untolerated taint {kwok.x-k8s.io/node: fake}. preemption: 0/1 nodes are available: 1 Preemption is not helpful for scheduling.
```

My needs:
1. Deploy volcano
2. Create some fake nodes, each node has different resources
3. Create some pods to be scheduled by volcano
4. pods are scheduled to those fake nodes according to volcano’s scheduling policy

**Comment by wzshiming** [comment, marked as answer]
kwok itself does not actually run the container [see more](https://kwok.sigs.k8s.io/docs/design/architecture)

I recommend using kwokctl to create a cluster without a scheduler, and then using kubeconfig to schedule the pods from an external scheduler.

```bash
kwokctl create cluster --disable-kube-scheduler

external-scheduler --kubeconfig ~/.kube/config
```

The following are cases for some projects
[Istio](https://kwok.sigs.k8s.io/docs/examples/istio/)
[Linkerd](https://kwok.sigs.k8s.io/docs/examples/linkerd/)
[Argo](https://kwok.sigs.k8s.io/docs/examples/argo/)
[KubeVirt](https://github.com/kwok-ci/kubevirt-demo)
[Kyverno](https://github.com/search?q=repo%3Akyverno%2Fkyverno%20kwok&type=code)
[YuniKorn](https://github.com/search?q=repo%3Aapache%2Fyunikorn-k8shim%20kwok&type=code)
[Karmada](https://karmada.io/docs/developers/performance-test-setup-for-karmada)
[Karpenter](https://github.com/search?q=repo%3Akubernetes-sigs%2Fkarpenter%20kwok&type=code)
[KubeAmiral](https://github.com/search?q=repo%3Akubewharf%2Fkubeadmiral%20kwok&type=code)
[MCAD](https://github.com/search?q=repo%3Aproject-codeflare%2Fmulti-cluster-app-dispatcher%20kwok&type=code)

**Reply by hwdef** [reply]
Thanks, I will try this
